accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,134 @@
+# Dart Package Configuration File v2.0
+
+Author: lrn@google.com
+Version: 0.1
+
+Dart currently uses a `.packages` file to configure *package URI resolution* (converting a package URI to a file name). The file uses a `.ini` file-like format with package names as keys and URI references as values. The URI references generated by Pub are always absolute or relative paths, so third-party users of the file tend to see them as paths, not URI.
+
+For the *language versioning* feature, we have proposed extending the format by adding fragments to the URI references and an entry with the empty string as key. This is a breaking change (the empty key was not previously accepted) and even more breaking for tools which assume the values are paths.
+
+Instead of breaking the existing file users, we will instead introduce a *new* file containing the information that we intended to store in the `.packages` file. This fill will use a (more) structured format which allows extending it with further information in the future. We will still generate the `.packages` file for now, allowing users of the file to continue working until they can migrate to using the new file.
+
+## New File Contents
+
+The file will contain the information that was planned for the `.packages` file:
+
+- A *default package* specified by its package name.
+- A list of *packages*, each with:
+  - A *package name*.
+  - A *path* or *location URI*.
+  - A *language version* (major/minor version number)
+
+On top of that, we also add extra metadata which allows us to version the format and attach information about when and how the file was created.
+
+We use JSON for the format. JSON has parsers available for most languages (not all Dart tools are written in Dart) and it is standardized and efficiently parsable.  JSON is *not* easy to write by hand (but doable), and it does not allow comments, but this is a machine generated file that will hardly ever be edited by hand.
+
+## New File Format
+
+The new file will be called `.packages.json` and stored in the same directory as the `.packages` file. The format is JSON text describing is a single JSON object.
```

Comment:
Just a naming nit to consider: 

`.packages.json` looks so similar to Node's `package.json`, so when both are in the same directory (which does happen) I can see how it can be very confusing.

I understand that this file is not specific to Pub so calling it `pubspec.packages` could be missleading too. But something like `dart.packages` (or `dart.packages.json`) would make it much more explicit.

Reply:
Naming is still open, but the name will definitely not be `packages.json`.